File: backend/src/main/java/com/auction/app/domains/auction/auction/Auction.java
package com.auction.app.domains.auction.auction;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Instant;

import com.auction.app.domains.products.Product;
import com.auction.app.domains.users.users.User;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.ColumnDefault;

@Entity
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "auctions")
public class Auction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id", nullable = false)
    private User seller;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = true)
    private Product product;

    @Column(nullable = false)
    private Integer auctionedQuantity;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal startingPrice;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal currentPrice;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal minBidIncrement;

    @Column(nullable = false)
    private Instant startTime;

    @Column(nullable = false)
    private Instant endTime;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private AuctionStatus status = AuctionStatus.UPCOMING;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "winner_id")
    private User winner;

    @Column(nullable = false)
    @Builder.Default
    private Integer bidCount = 0;

    @Column(nullable = false)
    @Builder.Default
    private boolean extended = false;

    public void recalculateMinBidIncrement() {
        this.minBidIncrement = this.currentPrice
                .multiply(BigDecimal.valueOf(0.05))
                .setScale(2, RoundingMode.HALF_UP);
    }
}

File: frontend/auction-app/src/types/auction.ts
import type { Tag } from './product';

export type AuctionStatus = 'UPCOMING' | 'ACTIVE' | 'ENDED' | 'CANCELLED';

export interface AuctionRequest {
    productId: number;
    quantity: number;
    startingPrice: number; // Đổi thành number để khớp với BigDecimal, hoặc string nếu bạn quản lý chuỗi lớn
    startTime: string;     // ISO String (Instant bên Java)
    endTime: string;       // ISO String (Instant bên Java)
}

export interface AuctionResponse {
    id: number;
    sellerId: number;
    sellerLabel: string;
    productId: number;
    productName: string;
    productTags: Tag[];
    productDescription?: string;
    productImageUrl?: string;
    auctionedQuantity: number;
    startingPrice: string;
    currentPrice: string;
    minBidIncrement: string;
    startTime: string;
    endTime: string;
    extended: boolean;
    status: AuctionStatus;
    winnerId: number | null;
    winnerLabel: string | null;
    bidCount: string;
}

export interface AuctionFindingRequest {
    productName?: string;
    tags?: Tag[]; // Khớp với Set<Tag> bên backend (truyền object hoặc xử lý mảng)
    startTime?: string;
    endTime?: string;
    minStartingPrice?: string;
    status?: AuctionStatus;
}